version: '3.8'

services:
  phpmyadmin:
    image: phpmyadmin
    container_name: phpmyadmin-go_news
    restart: always
    ports:
      - 8080:80
    environment:
      PMA_HOST: db
      PMA_PORT: ${DB_PORT}
      PMA_USER: 'root'
      PMA_PASSWORD: ${DB_PASSWORD_ROOT}
      MYSQL_ROOT_PASSWORD: ${DB_PASSWORD_ROOT}
  db:
    image: mysql:8-debian
    container_name: db-mysql-go_news
    restart: always

    healthcheck:
      test:
        [
          'CMD',
          'mysqladmin',
          'ping',
          '-h',
          'localhost',
          '-u',
          '$$MYSQL_USER',
          '-p $$MYSQL_PASSWORD',
        ]
      timeout: 20s
      retries: 10
    environment:
      MYSQL_USER: ${DB_USERNAME}
      MYSQL_PASSWORD: ${DB_PASSWORD}
      MYSQL_ROOT_PASSWORD: ${DB_PASSWORD_ROOT}
      MYSQL_DATABASE: ${DB_DATABASE}
    ports:
      - ${DB_PORT}:${DB_PORT}
    volumes:
      - ./db/mysql/backup:/var/lib/mysql
      - ./db/mysql/init:/docker-entrypoint-initdb.d

  app:
    container_name: go-news
    build: .
    env_file:
      - .env
    ports:
      - ${PORT}:${PORT}
    volumes:
      - ./:/usr/src/app
    restart: on-failure
    depends_on:
      db:
        condition: service_healthy
